### PR TITLE
prevent moving/dropping the pet egg if it's hatched

### PR DESCRIPTION
**Pull request by sader1992**
* **Addressed Issue(s)**: https://github.com/rathena/rathena/issues/4178

* **Server Mode**: both

* **Description of Pull Request**: 
the server is allowed moving the egg pet after hatched
to test this
```
@item 9003
@hatch 
@storall / @dropall
```

returning the pet to the egg afterword will fail

 
 [ x ] player to trade `trade_tradeadditem`
 [ x ] player to cart `pc_cart_additem`
 [ x ] player to sell `pc_can_sell_item`
 [ x ] player to storage `storage_canAddItem`
 [ x ] and droping the item `pc_candrop`
 [ ] ?

### CODE REVIEW

File: src/map/storage.cpp [hunk 1]
```diff
@@ -214,6 +214,9 @@ static enum e_storage_add storage_canAddItem(struct s_storage *stor, int idx, st
 
 	if (!stor->state.put)
 		return STORAGE_ADD_NOACCESS;
+	
+	if (items[idx].card[0] == CARD0_PET && items[idx].attribute == 1)
+		return STORAGE_ADD_INVALID;
```
Comment (aleos89) [inline src/map/storage.cpp]: If you could move this up above the `STORAGE_ADD_NOACCESS` check I would appreciate it!

**Comment by sader1992** [comment]
this not ready yet , it's the same problem with @dropall also i was thinking about WPE and trade
so i am thinking about making the egg bound to char , this would fix anything we didn't think about too

**Comment by sader1992** [comment]
also deleting the item using commands like delitem should not work on it

i was thinking the best idea would be is making a new bound element that bound the item to char and prevent anything from editing it

or we find a new way to make the egg usable like making a new sd item to save the egg value , i think this would be the **most reasonable solution**

File: src/map/pet.cpp [hunk 2]
```diff
@@ -913,18 +913,32 @@ static int pet_performance(struct map_session_data *sd, struct pet_data *pd)
...
+	//sd->inventory.u.items_inventory[i].attribute = 0;
+	
+	struct item it;
+	memset(&it, 0, sizeof(it));
+	it.nameid = pd->pet.egg_id;
+	it.identify = 1;
+	it.card[0] = CARD0_PET;
+	it.card[1] = GetWord(pd->pet.pet_id, 0);
+	it.card[2] = GetWord(pd->pet.pet_id, 1);
+	//need to check bound state for the item ?
```
Comment (sader1992) [inline src/map/pet.cpp]: here i was thinking what the best way to save the bound stat of the item

**Comment by sader1992** [comment]
this just how the idea of the code
ofc if it's a good idea , i need to retrieve the first commits/clean the code
and search for any other function that use the hidden egg

**Comment by sader1992** [comment]
i wonder if the way it's right now is coming from the official server

**Comment by Daegaladh** [comment]
> 
> 
> i wonder if the way it's right now is coming from the official server

On official servers, when you hatch an egg the item apparently is still there in your inventory, just hidden.
I've tested this by having 99 items + an egg, and I still wasn't able to get any more items after hatching it. Also, when you get the pet back to the egg, there's no "you got X egg" message.
Btw, on Aegis if you have overweight or your inventory full, you never get the items dropped on the floor in any case, I was planning on getting rid of all those `map_addflooritem`...

**Comment by Lemongrass3110** [comment]
Somehow the subject of this pull request changed into a strange direction. The behavior is 100% correct, only some places that might have not been taken into account need to be fixed. If you keep it like this(reverting to the old wrong Athena behavior), then we will have to reject and close this pull request and fix it ourselves. If you want to fix the issues, then please feel free to go on with it.

**Comment by sader1992** [comment]
sorry didn't know that the old code was like that xD
i just saw https://github.com/eathena/eathena/blob/0cb07145ea1431528d0591b6a8706856915b3782/src/map/pet.c#L297

so the correct thing to do is just retrieving to the start , fixing the problem of @dropall and @storeall
unfortunately i don't have the list of all the problem , i only seen the issue by Badarosk0 and the drop thing just came in mind
didn't think about it more as i thought if we just change the behavior any unknown issue would be fixed xD

so i guess this PR is ready like this

**Comment by sader1992** [comment]
@Lemongrass3110 i am not sure why (about pc_can_sell_item) but if it's something about WPE we might need to add to pc_cart_additem and trade_tradeadditem

**Comment by sader1992** [comment]
i just tested it ,and yes
so 
player2cart (no need to vending skill)
player2trade

**Comment by Lemongrass3110** [comment]
> i just tested it ,and yes
> so
> player2cart (no need to vending skill)
> player2trade
> player2mail
> 
> i was thinking of making a function that do the check , so any update later would be in one place , and it's better than copy paste the code all over the place
> maybe
> 
> ```c
> bool is_hatched_egg(struct item *item);
> 
> /*
>  return true if the item is hatched egg
>  return false if not
> */
> bool is_hatched_egg(struct item *item){
> 	if(item->card[0] == CARD0_PET && item->attribute == 1)
> 		return false;
> 
> 	return true;
> }
> 
> if(is_hatched_egg(item))
> 	return INVALID;
> ```
> 
> I don't know what the best place for it , the logical place , i think it's pet.cpp , however that mean adding the pet header to the other files

Probably either pet.hpp or itemdb.hpp
But I think that there might already be similar functions in the itemdb one. So maybe that one would be better. I was also thinking about this, when I changed it, but ended up being lazy and not wanting to change that many places.

File: src/map/itemdb.cpp [hunk 3]
```diff
@@ -573,6 +573,12 @@ bool itemdb_isrestricted(struct item* item, int gmlv, int gmlv2, bool (*func)(st
 	return true;
 }
 
+bool itemdb_ishatched_egg(struct item* item) {
+	if (item->card[0] == CARD0_PET && item->attribute == 1)
```
Comment (aleos89) [inline src/map/itemdb.cpp]: Either use nullpo to catch an invalid item or do like:
```suggestion
	if (item && item->card[0] == CARD0_PET && item->attribute == 1)
```

**Comment by admkakaroto** [comment]
Is it possible to block use the item Pet incubator if you already have the window hatching the eggs open?

**Pull request merged**